lib/logger_splunk_backend.ex

```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
 
   require Logger
 
-  @impl true
+  @impl :gen_event
   def init({__MODULE__, name}) do
     state = %{
       name: name,
       buffer: [],
       max_buffer: 32,
       buffer_size: 0,
-      connector: Logger.Backend.Splunk.Output.Http,
+      response: nil,
+      output: nil,
+      printing_response?: false,
+      error_device: :stdio,
+      erlang_host_json: ~s("host":""),
+      index_json: [],
       host: nil,
       level: :debug,
       format: @default_format,
       compiled_format: Logger.Formatter.compile(@default_format),
       metadata: [],
       token: ""
     }
+
     {:ok, configure([], state)}
   end
 
-  @impl true
-  def handle_call({:configure, opts}, state) do
-    {:ok, :ok, configure(opts, state)}
+  def init(__MODULE__) do
+    # use a default name
+    init({__MODULE__, __MODULE__})
   end
 
-  @impl true
-  def handle_call(:connector, %{connector: connector} = state) do
-    {:ok, {:ok, connector}, state}
+  @impl :gen_event
+  def handle_call({:configure, opts}, state) do
+    {:ok, :ok, configure(opts, state)}
   end
 
-  @impl true
-  def handle_event({level, _gl, {Logger, msg, ts, md}}, %{level: min_level} = state) do
-    state = if is_nil(min_level) or Logger.compare_levels(level, min_level) != :lt do
-      log_event(level, msg, ts, md, state)
-    else
+  @impl :gen_event
+  def handle_event({level, _gl, {Logger, msg, ts, md}}, state) do
+    %{level: min_level, buffer_size: buffer_size, max_buffer: max_buffer, response: response} =
       state
+
+    cond do
+      not meet_level?(level, min_level) ->
+        {:ok, state}
+
+      is_nil(response) ->
+        # we aren't in the middle of a request, so send the message right away
+        {:ok, log_event(level, msg, ts, md, state)}
+
+      buffer_size < max_buffer ->
+        # we are in the middle of a request, but there's still room to
+        # buffer. we'll send these messages when the current one completes,
+        # from handle_http_event/2.
+        {:ok, buffer_event(level, msg, ts, md, state)}
+
+      true ->
+        # we're in the middle of the request, but have reached our max buffer size
+        state = buffer_event(level, msg, ts, md, state)
+        {:ok, await_response(state)}
```

Comment:
I'm a bit confused on the state machine here. If we're in the middle of a request and the buffer is full (slow network), I would think we'd want to do something like just drop any incoming log events (or maybe `IO.puts` them). If I'm understanding `await_response/1` correctly, it blocks until it gets its HTTP messages. Could this potentially cause a user of the library, who's trying to `Logger.info(...)` in this situation, be blocked on that call?

Reply:
Yes, that's correct. The way it works is Logger sends this process a message for each log message. It also measures the length of the message queue, a proxy for unprocessed messages. If the message queue gets too long, at first Logger switches to sync mode, which doesn't return from `Logger.info/1` et al until the message is processed. If there are still too many messages, Logger will switch into discard mode where messages aren't sent to this process at all until the message queue drops far enough. The way we hook into that is by not processing more log messages when we've reached our buffer limit, so the messages queue up as messages in this process' queue.

Reply:
Oh, I see, interesting! I wasn't considering the interactions at the `Logger` level w.r.t. to async and discarding. That makes sense.